### ISSUE TITLE
Support for Other KNN Distance Metrics

### ISSUE THREAD

**Issue by jlevy44** [rusty1s/pytorch_cluster#26]
Could one compute the knn_graph using a cosine distance metric?

**Comment by rusty1s** [comment]
You need to adjust the distance computation [here](https://github.com/rusty1s/pytorch_cluster/blob/master/cuda/knn_kernel.cu#L34). We could add a flag to support different distance metrics. Feel free to submit a PR if you would like to contribute on this one.

**Comment by jlevy44** [comment]
Ok, will do. How does the knn computation introduced here compare to other ANN methods (say an approximated way to make this faster)?

**Comment by rusty1s** [comment]
What do you mean with ANN methods?

**Comment by jlevy44** [comment]
https://github.com/erikbern/ann-benchmarks

**Comment by rusty1s** [comment]
Ah I see, well for CPU computation we fallback to scipy. The GPU version does parallelization over the number of examples (each block) and number of nodes (threads), so it is only fast when using batches and when not having too many data points in each example (ideally around `1024`).

**Comment by jlevy44** [comment]
Ah makes sense. The data I am dealing with can range in size to order ~ 10k x 100k

GPU computed affinity does seem tractable if it can handle this size of data

**Comment by jlevy44** [comment]
I forked it. If the goal is just to form an unweighted adjacency, is the more sensible calculation ||a|| ||b|| - dot(a,b) rather than 1-dot(a,b)/(||a|| * ||b||), I suppose it doesn't matter, but I may implement the former, I see you've done that for euclidean.

**Comment by rusty1s** [comment]
Sounds good to me :)

**Comment by jlevy44** [comment]
Just a heads up, I've never coded in cuda before, and its been a bit of time since I've done any serious c++, so please forgive me if I botch some of this. ;)

**Comment by jlevy44** [comment]
For the non-cuda implementation, seems the cosine metric is not as used with KDTree, except given some transformation. We may have to use another indexing method.

**Comment by jlevy44** [comment]
I'll just replace the c implementation for now with sklearn's. It's slower, we can just use it as a placeholder as we make changes to the PR.

**Comment by liaopeiyuan** [comment]
> I forked it. If the goal is just to form an unweighted adjacency, is the more sensible calculation ||a|| ||b|| - dot(a,b) rather than 1-dot(a,b)/(||a|| * ||b||), I suppose it doesn't matter, but I may implement the former, I see you've done that for euclidean.

This is really genius! I was really confused at first, so it may be helpful to add a few lines of explanation in the source code.
